src/Microsoft.Health.Fhir.Core/Features/Operations/Import/ImportOrchestratorTask.cs

```diff
@@ -301,6 +316,135 @@ private async Task UpdateProgressAsync(ImportOrchestratorTaskContext context, Ca
             await _contextUpdater.UpdateContextAsync(JsonConvert.SerializeObject(context), cancellationToken);
         }
 
+        private async Task ExecuteImprotProcessingTaskAsync(CancellationToken cancellationToken)
+        {
+            _orchestratorTaskContext.TotalSizeInBytes = _orchestratorTaskContext.TotalSizeInBytes ?? 0;
+
+            foreach (var input in _orchestratorInputData.Input.Skip(_orchestratorTaskContext.CreatedTaskCount))
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();
+                }
+
+                while (_orchestratorTaskContext.RunningTaskIds.Count >= _orchestratorInputData.MaxConcurrentProcessingTaskCount)
+                {
+                    HashSet<string> completedTaskIds = new HashSet<string>();
+                    foreach (string taskId in _orchestratorTaskContext.RunningTaskIds)
+                    {
+                        TaskInfo latestTaskInfo = await _taskManager.GetTaskAsync(taskId, cancellationToken);
+
+                        if (latestTaskInfo.Status == TaskStatus.Completed)
+                        {
+                            CheckTaskResult(latestTaskInfo);
+                            completedTaskIds.Add(taskId);
+                        }
+                    }
+
+                    if (completedTaskIds.Count > 0)
+                    {
+                        _orchestratorTaskContext.RunningTaskIds.RemoveAll(id => completedTaskIds.Contains(id));
+                        await UpdateProgressAsync(_orchestratorTaskContext, cancellationToken);
+                    }
+
+                    await Task.Delay(TimeSpan.FromSeconds(PollingFrequencyInSeconds), cancellationToken);
+                }
+
+                (string processingTaskId, long endSequenceId, long blobSizeInBytes) = await CreateNewProcessingTaskAsync(input, cancellationToken);
+
+                _orchestratorTaskContext.RunningTaskIds.Add(processingTaskId);
+                _orchestratorTaskContext.CurrentSequenceId = endSequenceId;
+                _orchestratorTaskContext.TotalSizeInBytes += blobSizeInBytes;
+                _orchestratorTaskContext.CreatedTaskCount += 1;
+                await UpdateProgressAsync(_orchestratorTaskContext, cancellationToken);
+            }
+
+            while (_orchestratorTaskContext.RunningTaskIds.Count > 0)
+            {
+                HashSet<string> completedTaskIds = new HashSet<string>();
```

Comment:
Maybe we can wrap line 364 to line 382 as a method to avoid repeating them.

Reply:
thanks, updated.